backend/src/resolvers.js

```diff
@@ -12,78 +14,96 @@ export default {
   },
 
   Mutation: {
-    createPost: (parent, args, context) => {
-      // throw error if user does not exist
-      const user = context.dataSources.db.getUser(args.userId);
-      if (!user) {
-        throw new UserInputError('Invalid user', {
-          invalidArgs: 'Userid does not exist',
+    login: (parent, args, context) => {
+      const { email, password } = args;
+
+      if (password.length < 8) {
+        throw new UserInputError('400', {
+          invalidArgs:
+            '[Validation Errors] password is shorter then 8 characters',
         });
       }
 
-      const newPost = {
-        title: args.title,
-        authorid: args.userId,
-        votes: 0,
-      };
+      // get User
+      const user = context.dataSources.db.getUserByEmail(email);
 
-      return context.dataSources.db.createPost(newPost);
-    },
+      if (user == null) {
+        // throw not found
+        throw new UserInputError('404', {
+          invalidArgs: 'user not found',
+        });
+      }
+
+      // try to authorize
+      const canLogIn = bcrypt.compareSync(password, user.password);
 
-    upvotePost: (parent, args, context) => {
-      const voter = context.dataSources.db.getUser(args.userId);
-      if (!voter) {
-        throw new UserInputError('Invalid user', {
-          invalidArgs: 'Userid does not exist',
+      if (!canLogIn) {
+        // throw 401 or return null
+        throw new UserInputError('401', {
+          invalidArgs: 'Unauthorized',
         });
       }
+      return 'Bearer ' + createTokenFor(user);
+    },
 
-      const post = context.dataSources.db.getPost(args.id);
-      if (!post) {
-        throw new UserInputError('Invalid post', { invalidArgs: args.id });
+    signup: (parent, args, context) => {
+      const { name, email, password } = args;
+
+      if (password.length < 8) {
+        throw new UserInputError('400', {
+          invalidArgs:
+            '[Validation Errors] password is shorter then 8 characters',
+        });
       }
 
-      return context.dataSources.db.upvotePost(post.id, voter.id);
-    },
+      // try to find if the email already exist
+      const userExist = context.dataSources.db.getUserByEmail(email);
```

Comment:
Email is not validated. It is possible to use an empty string as email address. Possible fix is looking for the '@' symbol or doing a validation check.